signature/cose/conformance_test.go

```diff
@@ -0,0 +1,217 @@
+	UnprotectedHeaders *cborStruct `json:"unprotectedHeaders"`
+	Output             cborStruct  `json:"expectedOutput"`
+}
+
+type cborStruct struct {
+	CBORHex  string `json:"cborHex"`
+	CBORDiag string `json:"cborDiag"`
+}
+
+func TestConformance(t *testing.T) {
+	data, err := os.ReadFile("testdata/conformance.json")
+	if err != nil {
+		t.Fatalf("os.ReadFile() failed. Error = %s", err)
+	}
+	var sign1 sign1
+	err = json.Unmarshal(data, &sign1)
+	if err != nil {
+		t.Fatalf("json.Unmarshal() failed. Error = %s", err)
+	}
+	testSign(t, &sign1)
+	testVerify(t, &sign1)
+}
+
+// testSign does conformance check on COSE_Sign1_Tagged
+func testSign(t *testing.T, sign1 *sign1) {
+	signRequest, err := getSignReq(sign1)
+	if err != nil {
+		t.Fatalf("getSignReq() failed. Error = %s", err)
+	}
+	env := envelope{}
+	encoded, err := env.Sign(signRequest)
+	if err != nil || len(encoded) == 0 {
+		t.Fatalf("Sign() faild. Error = %s", err)
+	}
+	newMsg := generateSign1(env.base)
+	got, err := newMsg.MarshalCBOR()
+	if err != nil {
+		t.Fatalf("MarshalCBOR() faild. Error = %s", err)
+	}
+
+	// sign1.Output.CBORHex is a manually computed CBOR hex used as ground
+	// truth in the conformance test.
+	want := hexToBytes(sign1.Output.CBORHex)
+	if !bytes.Equal(want, got) {
+		t.Fatalf("unexpected output:\nwant: %x\n got: %x", want, got)
+	}
+
+	// Verify using the same envelope struct
+	// (Verify with UnmarshalCBOR is covered in the testVerify() part)
+	_, _, err = env.Verify()
+	if err != nil {
+		t.Fatalf("Verify() failed. Error = %s", err)
+	}
+}
+
+// testVerify does conformance check by decoding COSE_Sign1_Tagged object
+// into Sign1Message
+func testVerify(t *testing.T, sign1 *sign1) {
+	signRequest, err := getSignReq(sign1)
+	if err != nil {
+		t.Fatalf("getSignReq() failed. Error = %s", err)
+	}
+	env := envelope{}
+	encoded, err := env.Sign(signRequest)
+	if err != nil || len(encoded) == 0 {
+		t.Fatalf("Sign() faild. Error = %s", err)
+	}
+	//Verify after UnmarshalCBOR
+	var msg cose.Sign1Message
+	// sign1.Output.CBORHex is a manually computed CBOR hex used as ground
+	// truth in the conformance test.
+	if err := msg.UnmarshalCBOR(hexToBytes(sign1.Output.CBORHex)); err != nil {
+		t.Fatalf("msg.UnmarshalCBOR() failed. Error = %s", err)
+	}
+
+	// Not doing conformance check on CertChain and signature fields,
+	// becase every time we run this test, a new certChain and signer would be
+	// generated, and hence, a new signature would be generated after Sign().
+	// Instead, CertChain is verified in verifySignerInfo(), and signature is
+	// verified by go-cose's Verify() later on.
+	certs := []*x509.Certificate{testhelper.GetRSALeafCertificate().Cert, testhelper.GetRSARootCertificate().Cert}
+	certChain := make([]interface{}, len(certs))
+	for i, c := range certs {
+		certChain[i] = c.Raw
+	}
+	msg.Headers.Unprotected[cose.HeaderLabelX5Chain] = certChain
+	msg.Signature = env.base.Signature
+
+	newEnv := envelope{
+		base: &msg,
+	}
```

Comment:
not a strong opinion, I feel we can extract it out to a new submethod like `createNewEnv`?

Reply:
Sure. Will update the code